### PR TITLE
Download gems with threads

### PR DESCRIPTION
**Pull request by charliesome**
This pull request modifies `Gem::RequestSet` to use threads to download gems.

This significantly reduces the time it takes to download gems with many dependencies, especially on high latency connections.

Without this patch, installing Rails 4.0.0 on my home connection takes just over a minute.

When downloading gems in parallel with this patch, the total time taken to install Rails 4.0.0 drops to just 18 seconds.

This could be further reduced by installing gems in parallel, although more work would need to be done to ensure gems are not installed before their dependencies.

### CODE REVIEW

**Comment by djberg96** [comment]
+1 to getting this incorporated.

**Comment by moubry** [comment]
+1

File: lib/rubygems/config_file.rb [hunk 1]
```diff
@@ -427,6 +436,12 @@ def to_yaml # :nodoc:
                             DEFAULT_VERBOSITY
                           end
 
+    yaml_hash[:concurrent_downloads] = if @hash.key?(:concurrent_downloads)
```
Comment (segiddins) [inline lib/rubygems/config_file.rb]: `@hash.fetch(:concurrent_downloads, DEFAULT_CONCURRENT_DOWNLOADS)`

**Comment by indirect** [comment]
@charliesome can you rebase this? It seems good to merge after a rebase and the one comment is addressed.

I didn't see this when you opened it, but another option is to use Bundler to install rails, where you can already adjust the number of threads used to download and install gems. :)

**Comment by duckinator** [comment]
So what's the status on this? In https://github.com/rubygems/rubygems/pull/649#issuecomment-24590560, @charliesome said they restored `VerboseDownloadReporter`, but it still looks like it was removed when I look at the diff. Aside from that and @segiddins's [comment](https://github.com/rubygems/rubygems/pull/649/files#r49800359), this looks like it just needs a rebase and then a re-review.

**Comment by djberg96** [comment]
Calling @charliesome, please rebase.

**Comment by duckinator** [comment]
Given that the person who opened this PR has not responded to requests for rebasing it in 2.5 years, are there any objections to me taking over this? /cc @indirect 

Also, I know Charlie said they restored VerboseDownloadReporter (see https://github.com/rubygems/rubygems/pull/649#issuecomment-24590560), but the diff still shows that it was removed, so that would need to be added back.

**Comment by charliesome** [comment]
Heya, sorry about dropping the ball on this one. I won't have the time to finish this up.

👍 @duckinator if you'd like to take it over.

**Comment by duckinator** [comment]
No problem, it happens.

**Comment by duckinator** [comment]
So, a bit of a brain-dump as I'm working on this:

The only real issue is lib/rubygems/request_set.rb. I'm still in the process of figuring out how this code works.

Un-resolved merge: https://gist.github.com/duckinator/9f3fce74b4a6013e09263ebc3097098b#file-request_set-rb-L171-L203

<br>

Code before this PR: https://github.com/rubygems/rubygems/blob/2c36c1b1e6cea48d640ca3f7735c3f45b59f20f0/lib/rubygems/request_set.rb#L120-L137

Code in this PR: https://github.com/charliesome/rubygems/blob/cd7deb2666bd4ce5451b4dcdb4d311593a0b5260/lib/rubygems/request_set.rb#L125-L151

Current code: https://github.com/rubygems/rubygems/blob/master/lib/rubygems/request_set.rb#L156-L188

---

The first chunk is unchanged.

These snippets seem to be (roughly?) equivalent:

old:
```ruby
      path = req.download cache_dir

      inst = Gem::Installer.new path, options

      yield req, inst if block_given?
```

new:
```ruby
          req.spec.install options do |installer|
            yield req, installer if block_given?
          end
```

The big rescue block is simply for debug output.

*Problem:* The code in request_set.rb no longer has a separate step for downloading it; this now appears to be implicitly handled by `req.spec.install`. You can't parallelize the installs without introducing race conditions (afaik?), only downloads, so the fact that downloading is now (probably?) handled inside of `.install` complicates parallelizing downloads.

I'm having trouble figuring out what `req.spec` is referring to. I initially thought it was a `Gem::Specification`, but that doesn't have a `#install` method. So I can't figure out what the `install` method is actually doing.

**Comment by duckinator** [comment]
It looks like `req.spec` is a `Gem::Resolver::Specification` (see [lib/rubygems/resolver/specification.rb](https://github.com/rubygems/rubygems/blob/0749715e4bd9e7f0fb631a352ddc649574da91c1/lib/rubygems/resolver/specification.rb#L75-L98)).

As I suspected, that method downloads the the gem (see [lib/rubygems/resolver/specification.rb, line 91](https://github.com/rubygems/rubygems/blob/0749715e4bd9e7f0fb631a352ddc649574da91c1/lib/rubygems/resolver/specification.rb#L91)).

I'm not sure how to approach this. Possibly splitting `#install` into a `#download` and `#install`? But I don't know what else relies on `#install`, so I'm very hesitant to change its behavior.

**Comment by duckinator** [comment]
(A note since there's a lot of participants in this thread: If you don't want to see updates on this, there's an "Unsubscribe" button on the right side of the page.)

**Comment by indirect** [comment]
@duckinator I would probably try an approach of having the threaded downloader call `req.spec.source.download(spec, cache_dir)`. Then you could modify the `#install` method to check for the gem in `cache_dir` first, and use that file if it is there.

**Comment by duckinator** [comment]
@indirect 👍 sounds good to me.

**Comment by duckinator** [comment]
errr, why was this approved? this should not have been approved.

**Comment by duckinator** [comment]
Closing in favor of #1898.

**Pull request closed without merging**